[SDXL] Add blog post about Flax/JAX demo (#1544)
huggingface/blog: sdxl_jax.md

`jit=True` indicates that we want the pipeline call to be compiled. This will happen the first time we call `generate`, and it will be very slow – JAX needs to trace the operations, optimize them and convert to low-level primitives. We'll run a first generation to complete this process and warm things up:

```Python
start = time.time()
print(f"Compiling ...")
generate(default_prompt, default_neg_prompt)
print(f"Compiled in {time.time() - start}")
```

TODO: put the output time here

Once the code has been compiled, inference will be super fast. Let's try again!

```Python
start = time.time()
prompt = "llama in ancient Greece, oil on canvas"
neg_prompt = "cartoon, illustration, animation"
images = generate(prompt, neg_prompt)
print(f"Inference in {time.time() - start}")
```

TODO: put the output

## Benchmark

The following measures were obtained running SDXL 1.0 base for 20 steps, with the default Euler Discrete scheduler. We used Python 3.10 and jax version 0.4.16. These are the same specs used in our [demo Space](https://huggingface.co/spaces/google/sdxl).

|           | 4 images (1 per core) | 8 images (2 per core) |
|-----------|:---------------------:|:---------------------:|

> | TPU v5e-4 | 2.33                  | 4.99                  |

Author: ```suggestion
| TPU v5e-4 | 2.33                  | 4.99                  |
| A100 (PyTorch) | 2.26 | 4.17   |
```

Measurements use the most optimized setting we have for PyTorch / GPU:
- Flash Attention
- Torch Compile
- PyTorch 2.01 + CUDA 11.8

        neg_prompt_ids=neg_prompt_ids,
        guidance_scale=guidance_scale,
        jit=True,
    ).images

    # convert the images to PIL
    images = images.reshape((images.shape[0] * images.shape[1], ) + images.shape[-3:])
    return pipeline.numpy_to_pil(np.array(images))
```

`jit=True` indicates that we want the pipeline call to be compiled. This will happen the first time we call `generate`, and it will be very slow – JAX needs to trace the operations, optimize them and convert to low-level primitives. We'll run a first generation to complete this process and warm things up:

```Python
start = time.time()
print(f"Compiling ...")
generate(default_prompt, default_neg_prompt)
print(f"Compiled in {time.time() - start}")
```

This took about three minutes the first time we ran it.
But once the code has been compiled, inference will be super fast. Let's try again!

```Python
start = time.time()
prompt = "llama in ancient Greece, oil on canvas"
neg_prompt = "cartoon, illustration, animation"
images = generate(prompt, neg_prompt)
print(f"Inference in {time.time() - start}")
```

It now took about 2s to generate the 4 images!

## Benchmark

> The following measures were obtained running SDXL 1.0 base for 20 steps, with the default Euler Discrete scheduler. We compare Cloud TPU v5e with A100 for the same batch sizes. Do note that, due to parallelism, a TPU v5e-4 like the ones we use in our demo will generate **4 images** when using a batch size of 1 (or 8 images with a batch size of 2). Similarly, a TPU v5e-8 will generate 8 images when using a batch size of 1.

Reviewer: Is there any specific reason to benchmark on 20 iterations when the code has 25 iterations as default inference steps? Does the PyTorch version also use Euler Discrete scheduler?

Reviewer 2: The PyTorch version uses Euler Discrete too, that's determined by the [model configuration](https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/model_index.json#L8) and not the code.

We used 20 steps to keep the same settings the demo uses. That value seems to render fast with good quality, but 25 is also a very reasonable option. We can run some tests to see if quality improves noticeably with 25.

authors:
- user: pcuenq
- user: patrickvonplaten
---

# Stable Diffusion XL 🤝 JAX + TPUv5e

<!-- {blog_metadata} -->
<!-- {authors} -->

Stable Diffusion XL is a large image generation model, whose UNet component is about three times as large as the one in the previous version of the model. Deploying a model like this in production is challenging due to the increased memory requirements, as well as increased inference times.

Google Cloud TPUs are designed for parallel operation, which makes them very interesting for user-facing workloads. [Cloud TPU v5e](https://cloud.google.com/blog/products/compute/announcing-cloud-tpu-v5e-and-a3-gpus-in-ga) was recently announced: it offers almost the same computing power (in terms of TFLOPs) as its predecessor, while being significantly cheaper.

🧨 Diffusers JAX integration offers a convenient way to run SDXL on TPU via XLA, and we built a demo to showcase it. You can try it out in [this Space](https://huggingface.co/spaces/google/sdxl) or in the playground embedded below:

<script type="module" src="https://gradio.s3-us-west-2.amazonaws.com/3.37.0/gradio.js"> </script>
<gradio-app theme_mode="light" space="google/sdxl"></gradio-app>

> Under the hood, this demo runs on several TPU v5e-4 instances –each instance has 4 cores–, and takes advantage of parallelization to create 4 large 1024×1024 images in about 4 seconds, including all communications overhead!

Reviewer 3: Here, we have 4 images created in 4 seconds across 4 chips. Below, batch-size=1 generates the image in about 2.33s. Are we using different inference steps in the demo, or is it because of some other overhead (like front end)?

Reviewer 2: It's additional overhead, the backend returns the images in 3.02s (since the POST request is received until the response is sent, including generation and format conversions). The additional second should cover the communication roundtrip and UI processing. We'll try to clarify.

Image generation, however, is better suited for compilation. If we always generate the same number of images and they have the same size, then the output shapes are constant and known in advance. The text inputs _are also constant_: by design, Stable Diffusion and SDXL use fixed-shape embedding vectors (with padding) to represent the prompts typed by the user. Therefore, we can write JAX code that relies on fixed shapes and that can be greatly optimized!

- High-performance throughput for high batch sizes

Workloads can be scaled across multiple devices using JAX's [pmap](https://jax.readthedocs.io/en/latest/_autosummary/jax.pmap.html), which expresses single-program multiple-data (SPMD) programs. Applying pmap to a function will compile a function with XLA, then execute it in parallel on various XLA devices. 
For text-to-image generation workloads this means that increasing the number of images rendered simultaneously is straightforward to implement and doesn't compromise performance. For example, running SDXL on a TPU with 8 cores will generate 8 images in the same time it takes for 1 core to create a single image.

> TPUv5e chips come in multiple flavors, including 1, 4 and 8-core setups. This allows you to tailor the number of chips for your use case.

Reviewer 3: We don't use the term "core" with TPUv5e any more. I'd recommend "...8-chip..." instead here.

---
title: 🧨 Accelerating Stable Diffusion XL Inference with JAX on Cloud TPU v5e

> thumbnail: /blog/assets/sdxl-jax/thumbnail.png

Reviewer 4: This thumbnail might be too heavy, maybe try to compress it

Reviewer 2: You are right, I'll convert to jpg

authors:
- user: pcuenq
- user: jffacevedo
  guest: true
- user: alexspiridonov
  guest: true
- user: pmotter
  guest: true
- user: yyetim
  guest: true
- user: svaibhav
  guest: true
- user: vjsingh
  guest: true
- user: patrickvonplaten
---

# Accelerating Stable Diffusion XL Inference with JAX on Cloud TPU v5e

> <!-- {blog_metadata} -->
> <!-- {authors} -->

Reviewer 5: ```suggestion
```

from #1554

> Generative AI models, such as Stable Diffusion XL (SDXL), enable the creation of high-quality, realistic content with wide-ranging applications. However, harnessing the power of such models presents significant challenges and computational costs. SDXL is a large image generation model, whose UNet component is about three times as large as the one in the previous version of the model. Deploying a model like this in production is challenging due to the increased memory requirements, as well as increased inference times. Today, we are thrilled to announce that Hugging Face Diffusers now supports serving SDXL using JAX on Cloud TPUs, enabling high-performance, cost-efficient inference.

Reviewer 4: ```suggestion
Generative AI models, such as Stable Diffusion XL (SDXL), enable the creation of high-quality, realistic content with wide-ranging applications. However, harnessing the power of such models presents significant challenges and computational costs. SDXL is a large image generation model whose UNet component is about three times as large as the one in the previous version of the model. Deploying a model like this in production is challenging due to the increased memory requirements, as well as increased inference times. Today, we are thrilled to announce that Hugging Face Diffusers now supports serving SDXL using JAX on Cloud TPUs, enabling high-performance, cost-efficient inference.
```

🧨 Diffusers JAX integration offers a convenient way to run SDXL on TPU via [XLA](https://github.com/openxla/xla), and we built a demo to showcase it. You can try it out in [this Space](https://huggingface.co/spaces/google/sdxl) or in the playground embedded below:

<script type="module" src="https://gradio.s3-us-west-2.amazonaws.com/3.45.1/gradio.js"> </script>
<gradio-app theme_mode="light" space="google/sdxl"></gradio-app>

> Under the hood, this demo runs on several TPU v5e-4 instances (each instance has 4 TPU chips), and takes advantage of parallelization to serve 4 large 1024×1024 images in about 4 seconds. This time includes format conversions, communications time and frontend processing; the actual generation time is about 2.3s as we'll see below!

Reviewer 4: ```suggestion
Under the hood, this demo runs on several TPU v5e-4 instances (each instance has 4 TPU chips) and takes advantage of parallelization to serve four large 1024×1024 images in about 4 seconds. This time includes format conversions, communications time, and frontend processing; the actual generation time is about 2.3s, as we'll see below!
```

In this blog post,
1. [We describe why JAX + TPU + Diffusers is a powerful framework to run SDXL](#why-jax--tpuv5e-for-sdxl)
2. [Explain how you can write a simple image generation pipeline with Diffusers and JAX](#how-to-write-an-image-generation-pipeline-in-jax)
3. [Show benchmarks comparing different TPU settings](#benchmark)

## Why JAX + TPU v5e for SDXL?

Serving SDXL with JAX on Cloud TPU v5e with high performance and cost-efficiency is possible thanks to the application of combined hardware and software optimizations. Below we highlight two key factors: JAX just-in-time (jit) compilation and XLA compiler-driven parallelism with JAX pmap.

#### JIT compilation

> A notable feature of JAX is its [just-in-time (jit) compilation](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html). The JIT compiler traces code during the first run, and generates highly optimized TPU binaries that are re-used in subsequent calls.

Reviewer 4: ```suggestion
A notable feature of JAX is its [just-in-time (jit) compilation](https://jax.readthedocs.io/en/latest/jax-101/02-jitting.html). The JIT compiler traces code during the first run and generates highly optimized TPU binaries that are re-used in subsequent calls.
```

> The catch of this process is that it requires all input, intermediate and output shapes to be **static**, meaning that they must be known in advance. Every time we change the shapes

Reviewer 4: ```suggestion
The catch of this process is that it requires all input, intermediate, and output shapes to be **static**, meaning that they must be known in advance. Every time we change the shapes
```

> a new and costly compilation process will be triggered again. JIT compilation is ideal for services that can be designed around static shapes: compilation runs once and then we take advantage of super-fast inference times.

Reviewer 4: ```suggestion
a new and costly compilation process will be triggered again. JIT compilation is ideal for services that can be designed around static shapes: compilation runs once, and then we take advantage of super-fast inference times.
```

> Image generation is well-suited for JIT compilation. If we always generate the same number of images and they have the same size, then the output shapes are constant and known in advance. The text inputs are also constant: by design, Stable Diffusion and SDXL use fixed-shape embedding vectors (with padding) to represent the prompts typed by the user. Therefore, we can write JAX code that relies on fixed shapes and that can be greatly optimized!

Reviewer 4: ```suggestion
Image generation is well-suited for JIT compilation. If we always generate the same number of images and they have the same size, then the output shapes are constant and known in advance. The text inputs are also constant: by design, Stable Diffusion, and SDXL use fixed-shape embedding vectors (with padding) to represent the prompts typed by the user. Therefore, we can write JAX code that relies on fixed shapes, and that can be greatly optimized!
```

For text-to-image generation workloads this means that increasing the number of images rendered simultaneously is straightforward to implement and doesn't compromise performance. For example, running SDXL on a TPU with 8 chips will generate 8 images in the same time it takes for 1 chip to create a single image.

TPU v5e instances come in multiple shapes, including 1, 4 and 8-chip shapes, all the way up to 256 chips (a full TPU v5e pod), with ultra-fast ICI links between chips. This allows you to choose the TPU shape that best suits  your use case and easily take advantage of the parallelism that JAX and TPUs provide.

## How to write an image generation pipeline in JAX

We'll go step by step over the code you need to write to run inference super-fast using JAX! First, let's import the dependencies.

```Python
# Show best practices for SDXL JAX
import jax
import jax.numpy as jnp
import numpy as np
from flax.jax_utils import replicate
from diffusers import FlaxStableDiffusionXLPipeline
import time
```

> We'll now load the base SDXL model and the rest of components required for inference. The diffusers pipeline takes care of downloading and caching everything for us. Adhering to JAX's functional approach, the model's parameters are returned separately and will have to be passed to the pipeline during inference:

Reviewer 6: ```suggestion
We'll now load the base SDXL model and the rest of the components required for inference. The diffusers pipeline takes care of downloading and caching everything for us. Adhering to JAX's functional approach, the model's parameters are returned separately and will have to be passed to the pipeline during inference:
```

```Python
pipeline, params = FlaxStableDiffusionXLPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", split_head_dim=True
)
```

> Model parameters are downloaded in 32-bit precision by default. To save memory and run computation faster we'll convert them to `bfloat16`, an efficient 16-bit representation. However, there's a caveat: for best results we have to keep the _scheduler state_ in `float32`, otherwise precision errors accumulate and result in low-quality or even black images.

Reviewer 6: ```suggestion
Model parameters are downloaded in 32-bit precision by default. To save memory and run computation faster we'll convert them to `bfloat16`, an efficient 16-bit representation. However, there's a caveat: for best results, we have to keep the _scheduler state_ in `float32`, otherwise precision errors accumulate and result in low-quality or even black images.
```

# so we only need to replicate them once.
p_params = replicate(params)

def replicate_all(prompt_ids, neg_prompt_ids, seed):
    p_prompt_ids = replicate(prompt_ids)
    p_neg_prompt_ids = replicate(neg_prompt_ids)
    rng = jax.random.PRNGKey(seed)
    rng = jax.random.split(rng, NUM_DEVICES)
    return p_prompt_ids, p_neg_prompt_ids, rng
```

We are now ready to put everything together in a generate function:

```Python
def generate(
    prompt,
    negative_prompt,
    seed=default_seed,
    guidance_scale=default_guidance_scale,
    num_inference_steps=default_num_steps,
):
    prompt_ids, neg_prompt_ids = tokenize_prompt(prompt, negative_prompt)
    prompt_ids, neg_prompt_ids, rng = replicate_all(prompt_ids, neg_prompt_ids, seed)
    images = pipeline(
        prompt_ids,
        p_params,
        rng,
        num_inference_steps=num_inference_steps,
        neg_prompt_ids=neg_prompt_ids,
        guidance_scale=guidance_scale,
        jit=True,
    ).images

    # convert the images to PIL
    images = images.reshape((images.shape[0] * images.shape[1], ) + images.shape[-3:])
    return pipeline.numpy_to_pil(np.array(images))
```

> `jit=True` indicates that we want the pipeline call to be compiled. This will happen the first time we call `generate`, and it will be very slow – JAX needs to trace the operations, optimize them and convert to low-level primitives. We'll run a first generation to complete this process and warm things up:

Reviewer 6: ```suggestion
`jit=True` indicates that we want the pipeline call to be compiled. This will happen the first time we call `generate`, and it will be very slow – JAX needs to trace the operations, optimize them, and convert them to low-level primitives. We'll run a first generation to complete this process and warm things up:
```

The Cloud TPU tests were run using Python 3.10 and jax version 0.4.16. These are the same specs used in our [demo Space](https://huggingface.co/spaces/google/sdxl).

|                 | Batch Size | Latency  | Perf/$ |
|-----------------|------------|:--------:|--------|
| TPU v5e-4 (JAX) | 4          | 2.33s    | 21.46  |
|                 | 8          | 4.99s    | 20.04  |
| TPU v4-8 (JAX)  | 4          | 2.16s    | 9.05   |
|                 | 8          | 4.17     | 8.98   |

TPU v5e achieves 2.3 greater perf/$ on SDXL compared to TPU v4, demonstrating the cost-efficiency of the latest TPU generation.

## How does the demo work?

The [demo we showed before](https://huggingface.co/spaces/google/sdxl) was built using a script that essentially follows the code we posted in this blog post. It runs on a few Cloud TPU v5e devices with 4 chips each, and there's a simple load-balancing server that routes user requests to backend servers randomly. When you enter a prompt in the demo, your request will be assigned to one of the backend servers, and you'll receive the 4 images it generates.

> This is a simple solution based on a number of pre-allocated TPU instances. In a future post we'll cover how to create dynamic solutions that adapt to load using GKE.

Reviewer 6: ```suggestion
This is a simple solution based on several pre-allocated TPU instances. In a future post, we'll cover how to create dynamic solutions that adapt to load using GKE.
```
